app/uk/gov/hmrc/gform/playcomponents/GformSessionCookieCryptoFilter.scala

```diff
@@ -0,0 +1,184 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.gform.playcomponents
+
+import akka.stream.Materializer
+import play.api.Logger
+import play.api.http.HeaderNames
+import play.api.libs.crypto.CookieSigner
+import play.api.mvc.request.{ Cell, RequestAttrKey }
+import play.api.mvc._
+import cats.syntax.eq._
+import uk.gov.hmrc.crypto.{ Decrypter, Encrypter }
+import uk.gov.hmrc.gform.controllers.CookieNames
+import uk.gov.hmrc.gform.gformbackend.GformConnector
+import uk.gov.hmrc.gform.sharedmodel.formtemplate._
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.gform.controllers.CookieNames._
+import uk.gov.hmrc.play.bootstrap.frontend.filters.crypto.{ SessionCookieCrypto, SessionCookieCryptoFilter }
+import uk.gov.hmrc.gform.playcomponents.GformSessionCookieCryptoFilter._
+
+import scala.concurrent.{ ExecutionContext, Future }
+import scala.util.{ Failure, Success, Try }
+
+class GformSessionCookieCryptoFilter(
+  sessionCookieCrypto: SessionCookieCrypto,
+  gformConnector: GformConnector,
+  cookieSigner: CookieSigner,
+  val sessionBaker: SessionCookieBaker,
+  sessionBakerAnonymous: SessionCookieBaker
+)(
+  implicit
+  override val mat: Materializer,
+  override val ec: ExecutionContext)
+    extends SessionCookieCryptoFilter {
+
+  implicit val hc = HeaderCarrier()
+
+  override protected lazy val encrypter: Encrypter = sessionCookieCrypto.crypto
+  override protected lazy val decrypter: Decrypter = sessionCookieCrypto.crypto
+  override protected val decodeCookieHeader: String => Seq[Cookie] = Cookies.decodeCookieHeader
+
+  private val logger = Logger(getClass)
+
+  override def apply(next: (RequestHeader) => Future[Result])(rh: RequestHeader): Future[Result] =
+    decryptSession(rh).flatMap(dsResult => encryptSession(next(dsResult.rh), dsResult))
+
+  private def encryptSession(f: Future[Result], dsResult: DecryptSessionResult): Future[Result] =
+    f.map { result =>
+      val newSessionAsCookie: Option[Cookie] =
+        result.newSession.map { session =>
+          val sessionCookie: Cookie = dsResult.scBaker.encodeAsCookie(session)
+          sessionCookie.copy(value = encrypter.encrypt(sessionCookie.value))
+        }
+
+      val resultWithSessionReplacedByCookie: Option[Result] = {
+        newSessionAsCookie
+          .map { c =>
+            result
+              .copy(newSession = None)
+              .withCookies(c)
+          }
+      }
+
+      val cResult = resultWithSessionReplacedByCookie.getOrElse(result)
+
+      dsResult.fId.fold(cResult) { fid =>
+        cResult.withCookies(
+          Cookie(formTemplateIdCookieName, fid.value),
+          Cookie(formTemplateAuthConfigCookieName, dsResult.formTemplateAuthConfig.getOrElse(naFacCookieValue)))
+      }
+    }
+
+  private def decryptSession(rh: RequestHeader): Future[DecryptSessionResult] = {
+    val fIdInPath = Try(rh.path.split("/")(3)) match {
```

Comment:
This is really brittle. Cannot we extract `formTemplateId` from `app.routes` as play framework does?

Look at `target/scala-2.12/routes/main/app/Routes.scala`

They are pattern matching at `RequestHeader` and they have function taking `formTemplateId` inside:

```scala
def routes: PartialFunction[RequestHeader, Handler] = {
    ... 
    // @LINE:6
    case uk_gov_hmrc_gform_gform_NewFormController_dashboard2_route(params@_) =>
      call(params.fromPath[FormTemplateId]("formTemplateId", None)) { (formTemplateId) =>
        uk_gov_hmrc_gform_gform_NewFormController_dashboard2_invoker.call(NewFormController_7.dashboard(formTemplateId))
      }
    ...
}
```